62_CAD/DrawResourcesFiller.cpp

```diff
@@ -767,4 +809,132 @@ void DrawResourcesFiller::addHatch_Internal(const Hatch& hatch, uint32_t& curren
+	{
+		wchar_t k = wchar_t(text.at(i));
+		auto glyphIndex = m_face->getGlyphIndex(k);
+		const auto glyphMetrics = m_face->getGlyphMetrics(glyphIndex);
+		const float64_t2 baselineStart = currentBaselineStart;
+
+		// Vector transform
+		currentBaselineStart += mulMatrix3(transformation, float32_t3(glyphMetrics.advance, 0.0));
+
+		if (glyphIndex == 0 || (glyphMetrics.size.x == 0.0 && glyphMetrics.size.y == 0.0))
+			continue;
+
+		{
+			msdfgen::Shape shape = m_face->generateGlyphShape(glyphIndex);
+			_NBL_BREAK_IF(shape.contours.empty());
+		}
+
+		TextRenderer::GlyphBox glyphBbox = {
+			.topLeft = baselineStart + mulMatrix3(transformation, float32_t3(glyphMetrics.horizontalBearing.x, glyphMetrics.horizontalBearing.y, 0.0)),
+			.dirU = mulMatrix3(transformation, float64_t3(glyphMetrics.size.x, 0.0, 0.0)),
+			.dirV = mulMatrix3(transformation, float64_t3(0.0, -glyphMetrics.size.y, 0.0)),
+			.glyphIdx = glyphIndex,
+		};
+
+		glyphBoxes.push_back(glyphBbox);
+	}
+}
+
+void SingleLineText::Draw(TextRenderer* textRenderer, DrawResourcesFiller& drawResourcesFiller, SIntendedSubmitInfo& intendedNextSubmit)
+{
+	auto glyphBoxes = getGlyphBoxes();
+	for (auto glyphBox = glyphBoxes.data(); glyphBox < glyphBoxes.data() + glyphBoxes.size(); glyphBox++)
+	{
+		LineStyleInfo lineStyle = {};
+		lineStyle.color = float32_t4(1.0, 1.0, 1.0, 1.0);
+		const uint32_t styleIdx = drawResourcesFiller.addLineStyle_SubmitIfNeeded(lineStyle, intendedNextSubmit);
+
+		auto glyphObjectIdx = drawResourcesFiller.addMainObject_SubmitIfNeeded(styleIdx, intendedNextSubmit);
+
+		const auto msdfHash = hashFontGlyph(m_face->getHash(), glyphBox->glyphIdx);
+		DrawResourcesFiller::TextureReference* textureReference = drawResourcesFiller.addMSDFTexture(
+			[&]()
+			{
+				auto shape = m_face->generateGlyphUploadInfo(textRenderer, glyphBox->glyphIdx, uint32_t2(MSDFSize, MSDFSize));
+				MSDFTextureUploadInfo textureUploadInfo = {
+					.cpuBuffer = std::move(shape.msdfBitmap),
+					.bufferOffset = 0u,
+					.imageExtent = uint32_t3(MSDFSize, MSDFSize, 1),
+					.shapeSize = float32_t2(shape.biggerAxis == 1u ? shape.smallerSizeRatio : 1.0, shape.biggerAxis == 0u ? shape.smallerSizeRatio : 1.0),
+				};
+				return textureUploadInfo;
+			},
+			msdfHash,
+			intendedNextSubmit
+		);
+		const auto textureId = textureReference->alloc_idx;
+		assert(textureId != DrawResourcesFiller::InvalidTextureHash);
+
+		float minUVOffset = textRenderer->GetPixelRange() / MSDFSize;
+		float32_t2 aspectRatioUVOffset = ((float32_t2(1.0) - textureReference->shapeSize) * float32_t2(0.5)) 
+			* float32_t2(1.0 - 2.0 * minUVOffset);
+		float32_t2 minUV = float32_t2(minUVOffset) + aspectRatioUVOffset;
+		FontGlyphInfo glyphInfo = {
+			.topLeft = glyphBox->topLeft,
+			.dirU = glyphBox->dirU,
+			.aspectRatio = float(glyphBox->dirV.y / -glyphBox->dirU.x),
+			.minUV_textureID_packed = textureId | uint32_t(minUV.x * 255.0) << 24 | uint32_t(minUV.y * 255.0) << 16,
```

Comment:
make new pack/unpack funtions inside common.hlsl and use `nbl::hlsl::glsl::bitfieldInsert/Extract`